core/src/test/java/org/elasticsearch/indices/cluster/IndicesClusterStateServiceRandomUpdatesTests.java

```diff
@@ -97,7 +102,71 @@ public void testRandomClusterStateUpdates() {
         logger.info("Final cluster state: {}", state.prettyPrint());
     }
 
-    public ClusterState randomInitialClusterState(Map<DiscoveryNode, IndicesClusterStateService> clusterStateServiceMap) {
+    /**
+     * This test ensures that when a node joins a brand new cluster (different cluster UUID),
+     * different from the cluster it was previously a part of, the in-memory index data structures
+     * are all removed but the on disk contents of those indices remain so that they can later be
+     * imported as dangling indices.  Normally, the first cluster state update that the node
+     * receives from the new cluster would contain a cluster block that would cause all in-memory
+     * structures to be removed (see {@link IndicesClusterStateService#clusterChanged(ClusterChangedEvent)}),
+     * but in the case where the node joined and was a few cluster state updates behind, it would
+     * not have received the cluster block, in which case we still need to remove the in-memory
+     * structures while ensuring the data remains on disk.  This test executes this particular
+     * scenario.
+     */
+    public void testJoiningNewClusterOnlyRemovesInMemoryIndexStructures() {
+        // we have an IndicesClusterStateService per node in the cluster
+        final Map<DiscoveryNode, IndicesClusterStateService> clusterStateServiceMap = new HashMap<>();
+        ClusterState previousState = randomInitialClusterState(clusterStateServiceMap, RecordingIndicesService::new);
+        ClusterState newClusterState = ClusterState.builder(previousState)
+                                           .metaData(MetaData.builder(previousState.metaData()).clusterUUID(UUIDs.randomBase64UUID()))
+                                           .build();
+
+        // add index to the cluster
+        String name = "index_" + randomAsciiOfLength(8).toLowerCase(Locale.ROOT);
+        CreateIndexRequest request = new CreateIndexRequest(name, Settings.builder()
+                                                                      .put(SETTING_NUMBER_OF_SHARDS, randomIntBetween(1, 3))
+                                                                      .put(SETTING_NUMBER_OF_REPLICAS, 1)
+                                                                      .build());
+        ClusterState state = cluster.createIndex(previousState, request);
+        assertTrue(state.metaData().hasIndex(name));
+
+        // pick a data node to simulate the cluster state change on
+        DiscoveryNode node = null;
+        for (DiscoveryNode discoveryNode : clusterStateServiceMap.keySet()) {
+            if (discoveryNode.isDataNode()) {
```

Comment:
maybe you should pick a node that has actually a shard of the index assigned to it. This allows to check that the node actually had in-memory structure of the index before it was removed.
```
DiscoveryNode nodeWithShard = state.nodes().get(randomFrom(state.routingTable().index(name).shardsWithState(INITIALIZING)).currentNodeId());
```